airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/cursor.py

```diff
@@ -143,6 +162,7 @@ def __init__(
         end_provider: Callable[[], CursorValueType],
         lookback_window: Optional[GapType] = None,
         slice_range: Optional[GapType] = None,
+        cursor_granularity: Optional[GapType] = None,
```

Comment:
`cursor_granularity` is a part of low-code cursors, but not concurrent. The main lever this turns is the reduction of record duplicates from date window edges the overlap.

Our overall stance has been that some duplicates over the course of a sync are tolerable, but we should never lose records. After auditing our existing repository connectors, we have 14 cases where we use "P1D" or greater as granularity. This I would consider to be a potentially unacceptable number of duplicates and this is on some of our more highly used connectors hence why I've added this in

Reply:
thanks for checking how often the feature is used

Reply:
I like the change of supporting granularity here.

For the topic of duplication and big granularities, isn't it driven by the API granularity? For example, if an API only takes format `YYYY-MM` as an input, I don't see how we can have something different than `P1M` as a cursor granularity

Reply:
At the largest sizes we do start to lose some options. Although I don't think the month or year option is that common or if I've ever seen it. Bad API design in my opinion. I think there's an interesting question of inference in that it's probably 99% chance a user using a date time of hours would use hour granularity. But that might be more of low-code interface as opposed to the underlying cursor. Which we would still need `cursor_granularity` configurable